ksmt-core/src/main/kotlin/org/ksmt/utils/FpUtils.kt

```diff
@@ -0,0 +1,1430 @@
+         * */
+        val multipliedSignificand = lhs.significand * rhs.significand
+
+        /**
+         *  Remove the extra bits, keeping a sticky bit.
+         *  Multiplication result is of the form:
+         *  [significandSize result bits][4 special bits][significandSize-4 extra bits]
+         *  */
+        var (normalizedSignificand, stickyRem) = if (lhs.significandSize >= 4u) {
+            val resultWithReminder = multipliedSignificand.divideAndRemainder(
+                powerOfTwo(lhs.significandSize - 4u)
+            )
+            resultWithReminder[0] to resultWithReminder[1]
+        } else {
+            // Ensure significand has at least 4 bits (required for rounding)
+            val correctedSignificand = multipliedSignificand.mul2k(4u - lhs.significandSize)
+            correctedSignificand to BigInteger.ZERO
+        }
+
+        if (!stickyRem.isZero() && normalizedSignificand.isEven()){
+           normalizedSignificand++
+        }
+
+        val unpackedResult = UnpackedFp(
+            lhs.exponentSize, lhs.significandSize, resultSign, resultExponent, normalizedSignificand
+        )
+
+        return fpRound(rm, unpackedResult)
+    }
+
+    private fun KContext.fpDivUnpacked(rm: KFpRoundingMode, lhs: UnpackedFp, rhs: UnpackedFp): KFpValue<*> {
+        val resultSign = lhs.sign xor rhs.sign
+        val resultExponent = lhs.unbiasedExponent - rhs.unbiasedExponent
+
+        val extraBits = lhs.significandSize + 2u
+        val lhsSignificandWithExtraBits = lhs.significand.mul2k(lhs.significandSize + extraBits)
+
+        /**
+         * Since significands are normalized (both have MSB 1),
+         * the division result will have at least [extraBits] bits.
+         * */
+        val divisionResultSignificand = lhsSignificandWithExtraBits.divide(rhs.significand)
+
+        /**
+         *  Remove the extra bits, keeping a sticky bit.
+         *  [normalizedResultSignificand] will have at least 4 bits as required for rounding.
+         *  */
+        var (normalizedResultSignificand, stickyRem) = divisionResultSignificand.divideAndRemainder(
+            powerOfTwo(lhs.significandSize)
+        )
+        if (!stickyRem.isZero() && normalizedResultSignificand.isEven()) {
+            normalizedResultSignificand++
+        }
+
+        val unpackedResult = UnpackedFp(
+            lhs.exponentSize, lhs.significandSize, resultSign, resultExponent, normalizedResultSignificand
+        )
+
+        return fpRound(rm, unpackedResult)
+    }
+
+    private fun KContext.fpSqrtUnpacked(rm: KFpRoundingMode, value: UnpackedFp): KFpValue<*> {
+        /**
+         * Add significandSize + 6 bits, to ensure that sqrt
+         * will have at least significandSize + 3 bits.
+         *
+         * Add one more (7) bits in case of even exponent to get more precise result.
+         * This extra bit is handled by decrementing the [resultExponent] by one.
+         * */
```

Comment:
Could you add comments about the source for these methods, just like for the addition implementation? It is quite hard to understand, why it should be `6` here